File: android/src/main/kotlin/com/example/ptnet_plugin/PtnetPlugin.kt
package com.example.ptnet_plugin

import android.app.Activity
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


/** PtnetPlugin */
class PtnetPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ptnet_plugin")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        // Get data from dart
        val map: Map<String, Any?> = call.arguments as Map<String, Any?>

        val ttl = map.getOrDefault("ttl", -1) as Int
        val port = map.getOrDefault("port", -1) as Int
        val timeout = map.getOrDefault("timeout", -1) as Int
        val server = map.getOrDefault("server", "") as String
        val address = map.getOrDefault("address", "") as String

        // Note: Lỗi ở Ping, reset máy thử ( recommend: ko dùng AVD-Android Virtual Device (AndroidStudio-setting cực)
        val librariesHandler = LibrariesHandler()
        when (call.method) {
            "getPingResult" -> result.success(
                librariesHandler.pingResult(address)
            )

            "getPageLoadResult" -> result.success(
                librariesHandler.pageLoadResult(address)
            )

            "getDnsLookupResult" -> result.success(
                librariesHandler.dnsLookUpResult(address, server)
            )

            "getPortScanResult" -> result.success(
                librariesHandler.portScanResult(address, port, timeout)
            )

            "getTraceRouteResult" -> result.success(
                librariesHandler.traceRouteResult(address, ttl)
            )

            "getTraceRouteEndpoint" -> result.success(
                librariesHandler.traceRouteEndpoint(address)
            )

            "getWifiScanResult" -> result.success(
                librariesHandler.wifiScanResult()
            )

            "getWifiInfo" -> result.success(
                librariesHandler.wifiInfo()
            )

            else -> result.notImplemented()
        }
    }

    //-----------------------------------------------------------------------------
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null;
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null;
    }
}

File: ios/Classes/PtnetPlugin.swift
import Flutter
import PTNetFramework

public class PtnetPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "ptnet_plugin", binaryMessenger: registrar.messenger())
    let instance = PtnetPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let libraryHandler = LibraryHandler()

    switch call.method {
    case "getPlatformVersion":
      libraryHandler.getPlatformVersion { platformVersion in
        result(platformVersion)
      }
    case "getWifiInfo":
      libraryHandler.getWifiInfo { wifiInfo in
        result(wifiInfo)
      }
    case "getPingResult":
      if let arguments = call.arguments as? [String: Any],
         let address = arguments["address"] as? String {
        libraryHandler.pingAddress(address: address) { pingResult in
          result(pingResult)
        }
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing address argument", details: nil))
      }
    case "getPageLoadResult":
      if let arguments = call.arguments as? [String: Any],
         let address = arguments["address"] as? String {
        libraryHandler.loadPage(address: address) { pageLoadResult in
          result(pageLoadResult)
        }
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing address argument", details: nil))
      }
    case "getDnsLookupResult":
      if let arguments = call.arguments as? [String: Any],
         let domainName = arguments["address"] as? String,
         let dnsServer = arguments["server"] as? String {
        libraryHandler.resolveDomain(domainName: domainName, dnsServer: dnsServer) { dnsLookupResult in
          result(dnsLookupResult)
        }
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing domainName or dnsServer argument", details: nil))
      }
    case "getPortScanResult":
      if let arguments = call.arguments as? [String: Any],
         let address = arguments["address"] as? String,
         let port = arguments["port"] as? Int,
         let timeout = arguments["timeout"] as? Double {
        libraryHandler.scanPort(address: address, port: port, timeout: timeout) { portScanResult in
          result(portScanResult)
        }
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing address, port, or timeout argument", details: nil))
      }
    case "getTraceRouteResult":
      if let arguments = call.arguments as? [String: Any],
         let host = arguments["address"] as? String,
         let ttl = arguments["ttl"] as? Int {
        libraryHandler.traceRoute(host: host, ttl: ttl) { traceRouteResult in
          result(traceRouteResult)
        }
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing host or ttl argument", details: nil))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
